### PR TITLE
Implement a basic swarm with an `echo` example

### PR DESCRIPTION
**Pull request by tomaka**
Based over #40.

What this PR does is:

- Remove the `host` and `transport` library, which were just stubs anyway.
- Introduce the `swarm` library, whose purpose is to contain all the traits and architecture to plug things together.
- `swarm` contains the `Transport` trait, which is implemented on `Tcp` (within `libp2p-tcp-transport`) and on a dummy `DeniedTransport` struct, and the `ConnectionUpgrade` trait, which is implemented on `SecioConnUpgrade` (within `libp2p-secio`) and on `PlainText`.
- The `Transport` and `ConnectionUpgrade` traits have methods that make it possible to combine them into a chain (similar to futures, iterators, and so on). You can therefore build powerful chains such as "use UDP only with secio, or TCP with plaintext or secio".
- Adds an `example` crate in the workspace that contains the `echo-dialer` and `echo-listener` examples. They can interoperate with the `echo` example [found here](https://github.com/libp2p/js-libp2p/tree/master/examples/echo).
- Modifies `Tcp::new` to take a tokio `Handle` instead of build its own tokio reactor, which makes it possible to use the same tokio reactor for both TCP and UDP for example.

### CODE REVIEW

File: example/examples/echo-dialer.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER 
+// DEALINGS IN THE SOFTWARE.
+
+extern crate bytes;
+extern crate futures;
+//extern crate libp2p_secio as secio;
```
Comment (rphmeier) [inline example/examples/echo-dialer.rs]: can you remove commented-out code in this file?

File: libp2p-secio/src/lib.rs [hunk 2]
```diff
@@ -54,6 +59,50 @@ mod keys_proto;
 mod handshake;
 mod structs_proto;
 
+#[derive(Clone)]
+pub struct SecioConnUpgrade {
```
Comment (rphmeier) [inline libp2p-secio/src/lib.rs]: missing docs

File: libp2p-swarm/src/transport.rs [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+					upgrade: upgrade,
+				};
+
+				return Err((builder, addr));
+			}
+		};
+
+		let future = dialed_fut
+            // Try to negotiate the protocol.
+            .and_then(move |connection| {
```
Comment (rphmeier) [inline libp2p-swarm/src/transport.rs]: indentation here is a little messed up

**Comment by tomaka** [comment]
Additional comment: I took a very conservative (and therefore less controversial?) approach for the API. In other words this API should be extensible in any way we want. For example if later we want to be able to add new transports at runtime, we can just create a new struct that implements `Transport` and that allows adding instances of transport at runtime.

**Review by rphmeier — CHANGES_REQUESTED** [review]
looks good in general. some issues with indentation around `UpgradedNode`.

`OrTransport` dialing seems a little naive with the linear search. I know this maps the best to the type system, but some kind of routable hinting mechanism on dialer is worth considering in the future in the case that there are lots of transports.

File: example/examples/echo-dialer.rs (hunk 1 above)
Comment (tomaka) [reply example/examples/echo-dialer.rs]: I enabled `secio` instead, which is what I meant to do.

**Review by rphmeier — APPROVED** [review]

**Pull request merged**